src/Model/BSONIterator.php

```diff
@@ -0,0 +1,117 @@
+<?php
+/*
+ * Copyright 2018 MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace MongoDB\Model;
+
+use MongoDB\BSON;
+use MongoDB\Exception\UnexpectedValueException;
+use MongoDB\Model\BSONDocument;
+use Iterator;
+
+/**
+ * Iterator for BSON documents.
+ */
+class BSONIterator implements Iterator
+{
+    private $buffer;
+    private $bufferLength;
+    private $current;
+    private $key = 0;
+    private $position = 0;
+
+    const BSON_SIZE = 4;
+
+    /**
+     * Constructor.
+    *
+     * @internal
+     * @param string $data Concatenated, valid, BSON-encoded documents
+     */
+    public function __construct($data)
+    {
+        $this->buffer = $data;
+        $this->bufferLength = strlen($data);
+    }
+
+    /**
+     */
+    public function current()
+    {
+        return $this->current;
+    }
+
+    /**
+     * @see http://php.net/iterator.key
+     * @return mixed
+     */
+    public function key()
+    {
+        return $this->key;
+    }
+
+    /**
+     * @see http://php.net/iterator.next
+     * @return void
+     */
+    public function next()
+    {
+        $this->key++;
+        $this->current = null;
+        $this->getNextBSON();
+    }
+
+    /**
+     * @see http://php.net/iterator.rewind
+     * @return void
+     */
+    public function rewind()
+    {
+        $this->key = 0;
+        $this->position = 0;
+        $this->current = null;
+        $this->getNextBSON();
+    }
+
+    /**
+     * @see http://php.net/iterator.valid
+     * @return boolean
+     */
+    public function valid()
+    {
+        return $this->current !== null;
+    }
+
+    private function getNextBSON()
+    {
+        if ($this->position === $this->bufferLength) {
+            return;
+        }
+
+        if (($this->bufferLength - $this->position) < 5) {
+            throw new UnexpectedValueException(sprintf('Expected at least %d bytes; %d remaining', self::BSON_SIZE, $this->bufferLength - $this->position));
+        }
+
+        list(,$documentLength) = unpack('V', substr($this->buffer, $this->position, self::BSON_SIZE));
+
+        if (($this->bufferLength - $this->position) < $documentLength) {
+            throw new UnexpectedValueException(sprintf('Expected %d bytes; %d remaining', $documentLength, $this->bufferLength - $this->position));
+        }
+
+        $this->current = new BSONDocument((array)(substr($this->buffer, $this->position, self::BSON_SIZE + $documentLength)));
```

Comment:
This doesn't look correct, as raw bytes from BSON data are being cast to an array and fed to the BSONDocument constructor. Try to `var_dump()` the elements of this iterator to see what this produces.

You're going to need to use the [BSON functions](http://php.net/manual/en/ref.bson.functions.php) to convert these bytes into a PHP value to assign to `$this->current`. Additionally, since `MongoDB\BSON\toPHP()` supports a type map it'd be reasonable if the constructor accepted an optional type map. In the interest of consistency, it should probably be through a new `array $options = []` argument and parsed similar type maps for operation classes.

If there is a type map available, we can use it when calling `toPHP()`; otherwise, we can default to an empty array, which is equivalent to no type map for `toPHP()`. I'm not sure if `null` is accepted and ignored in the same fashion, but you could check the extension source to confirm.